### DISCUSSION TITLE
Quad dominant or all Quad mesh in 2D (preserving grain boundaries)

### DISCUSSION THREAD

**Discussion by MiguelVCarvalho** [neperfepx/neper#98] [Q&A] [answered]
Include the possibility of having **Quad dominant** meshes or **All Quads** meshes in 2D.

For example with the options:
`-elttype quad_dominant`
`-elttype all_quad`

**Why:** If studying intergranular fracture with cohesive elements, the grain geometry is of utmost importance. Currently, only triangular elements preserve the grain geometry, since with the option `-elttype quad`, a regular grid is created.
However, if the grains are modelled with an incompressible material (as it is the case of crystal plasticity), the standard tri 3 and tri 6 elements should not be used since they are highly prone to volumetric locking.
In this case, quad 4 (FBAR for example) or quad 8 elements should be used.

**Case 1:**
Gmsh enables the user to choose "Recombine all triangular meshes" while meshing a geometry. This will lead to a quad dominant mesh.
Example:
![image](https://user-images.githubusercontent.com/25244490/74353802-57285580-4db2-11ea-82f5-f289405ae513.png)
![image](https://user-images.githubusercontent.com/25244490/74353826-61e2ea80-4db2-11ea-8dd6-3b57bfcc0a30.png)

(in this case, it was able to generate only quad elements but sometimes, due to complex geometry, some triangular elements remain)

**Case 2:**
Gmsh enables the user to change the Subdivision algorithm to "All Quads" while meshing a geometry. This will lead to a mesh that only contains quad elements (the element size may reduce).
Example:
![image](https://user-images.githubusercontent.com/25244490/74354049-b38b7500-4db2-11ea-82ce-a5f3671df625.png)
![image](https://user-images.githubusercontent.com/25244490/74354067-b9815600-4db2-11ea-8730-c8106f8fe570.png)

**Disclaimer:** These examples may not present the best meshes in terms of element quality but it was something done quite fast and only for demonstration purposes.

**Comment by MiguelVCarvalho** [comment]
To use the subdivision algorithm **All Quads**, the checkbox **Recombine all triangular meshes** has to be activated also. Otherwise, it will only break all triangles into 4 quads.

**Reply by Mahdieh-karami** [reply]
Hi, Could you please explain more. I have a .msh4 file from Neper and I could open it in Gmsh. Although I changed mesh to all qauds, it seems that changing the mesh is not possible.

**Reply by rquey** [reply]
This is more a Gmsh issue. I don't know more than what you can find in this thread.

**Reply by Mahdieh-karami** [reply]
![image](https://github.com/user-attachments/assets/b0e25153-45c1-4ed8-a6fc-d04427b9e2ac)
Hi . Finally I could reach this mesh  Is it possible to change the size of elements ?I followed this :
neper -T -n 4 -dim 2
neper -M n4-id1.tess -format msh4
gmsh -2 n4-quad.geo 

and in .geo file :
Merge "n4-id1.msh4";
Mesh.RecombinationAlgorithm = 3;
RecombineMesh;
Mesh.SecondOrderLinear = 1;
Mesh.SubdivisionAlgorithm = 1;
RefineMesh;

**Reply by rquey** [reply]
Yes, use [`-rcl` (or `-cl`)](https://neper.info/doc/neper_m.html#cmdoption-cl-ver-edge-face) in -M.

**Reply by Mahdieh-karami** [reply]
Thanks. It worked, can I use same method in 3D ? I have problem with creating .msh4 file in 3D.

**Reply by rquey** [reply]
[`-rcl` (or `-cl`)](https://neper.info/doc/neper_m.html#cmdoption-cl-ver-edge-face) applies both to 2D and 3D.

**Reply by Mahdieh-karami** [reply]
Sorry for asking too many questions but I faced this error for both  .msh and .msh4  :
![image](https://github.com/user-attachments/assets/99a4a782-e94f-42f2-9ae5-b9efceedf95b)

**Reply by rquey** [reply]
Update your Neper.

**Reply by Mahdieh-karami** [reply]
It worked, thanks. what about .geo file  for 3D? Because the last .geo does not work on 3D.

**Reply by rquey** [reply]
It should... Maybe you need to adapt it, but a geo file works in 3D too.

**Reply by Mahdieh-karami** [reply]
Hi, 
I tried to adapt it . I faced this errors but it gives a Msh file in which there are just finer Tet elements .
![image](https://github.com/user-attachments/assets/b9b249c9-81cd-44c6-83b3-d81700456062)

**Comment by rquey** [comment]
How would it translate in terms of entries in the .geo file? For tri/tet meshing, Neper uses this:
```
/* Options for the meshing: */
[...]
Mesh.Algorithm      = 6;
Mesh.Algorithm3D    = 4;
[...]
Mesh.SaveAll = 1;
Mesh.Optimize       = 1;
```

**Comment by MiguelVCarvalho** [comment]
Quad Dominant:

`Mesh.RecombinationAlgorithm = 3;` (blossom full-quad)
`Mesh.RecombineAll                  = 1;`

All Quads:
`Mesh.RecombinationAlgorithm = 3;` (blossom full-quad)
`Mesh.RecombineAll                  = 1;`
`Mesh.SubdivisionAlgorithm       = 1;` (All quads)

Something like this might work.

**Comment by rquey** [comment, marked as answer]
@MiguelVCarvalho - As of version 3.5.3-29, a mesh can be written under format msh version 4, using `-format msh4`. The mesh file can then be processed externally (by Gmsh) to get a full-quad mesh.

Here is an example:
```
neper -T -n 2 -dim 2
neper -M n2-id1.tess -format msh4
gmsh -2 n2-quad.geo
```

This is only a start - it will be integrated to Neper more thoroughly.

n2-quad.geo:
```
Merge "n2-id1.msh4";
Mesh.RecombinationAlgorithm = 1;
RecombineMesh;
Mesh.SecondOrderLinear = 1;
Mesh.SubdivisionAlgorithm = 1;
RefineMesh;
```

**Reply by MiguelVCarvalho** [reply]
@rquey sorry for the delay in my response. I had some work regarding other subjects and I am now returning to the topic at hand.

Neper is currently in version 4.2.0. Just wanted to know if quad-dominant meshes have been implemented more thoroughly or if I should do as you have told above (create in msh version 4 and process it externally by Gmsh).

Thank you for your work!

**Reply by rquey** [reply]
I haven't done it more thoroughly yet, for good reasons it seems ;-). The solution above should still work.

**Reply by MiguelVCarvalho** [reply]
Hi @rquey , are you thinking of implementing this quad-dominant meshes more thoroughly in neper any time soon?

The solution above works more or less. I am creating the .geo file with `neper -T` and then using gmsh to create the `.msh` file. However, with periodic tesselations, periodic meshes are not obtained because using this external processing approach, the periodic information is missing and gmsh does not create periodic finite elements.

**Reply by rquey** [reply]
Hi @MiguelVCarvalho - Unfortunately no. Not high in the todo list...

**Reply by MiguelVCarvalho** [reply]
Thank you anyway @rquey ;)

Just out of curiosity, what type of element do you use when modelling crystal plasticity?

**Reply by rquey** [reply]
[10-node tets](https://fepx.info) ;)

**Comment by MiguelVCarvalho** [comment]
Don't you start having convergence problems (at the global equilibrium procedure) when the plastic deformation (incompressible) becomes dominant?

I have those problems for the tri6 and tet10.

(sorry, I know that this is not the place for this type of discussion)

**Reply by rquey** [reply]
@mkasemer could elaborate on this point, but no, we can go to large plastic strain (30-50%) no problem...

**Reply by mkasemer** [reply]
In general @rquey is correct: the proof is in the pudding, so to speak: the 10-node tets are demonstratively adept at modeling large deformation plasticity (let alone when plastic deformation becomes "dominant", which would imply 3-5% deformation for most metals). It is dependent on morphology of the microstructure and material system (and thus local mesh quality), but again @rquey is correct that FEPX can handle out to 30-50% deformation without the aid of re-meshing.

Regarding an historical perspective, I would not be the best to ask about this, but Paul Dawson would often cite early (1990s) issues when experimenting with 4-node linear tetrahedra---specifically shear locking phenomena---as the impetus to move to quadratic tets (10-node). The added degrees of freedom and ability to have non-linear shapes allowed for greatly reduced numerical over-stiffening (problems with 4-node tets), according to Dawson. Again, I have no first-hand experience with this.

Others have been moving to partial employ of brick elements recently, though I am skeptical of the absolute need for this.

Another question worth asking is: what quadrature rule are you using? FEPX's quadrature rules can be found in the source code. This could also influence stability of the solution.

**Reply by rquey** [reply]
As a quick follow-up to @mkasemer's response, Paul Dawson (@dplab) would also say that quad elements are less apt to large deformation - they generally diverge sooner. I think that FEPX had support for quad elements in the 90s - early 2000s, but it was then abandoned.

**Comment by dplab** [comment]
Hi Romain,
   Both statements are correct.
paul
ps  I've been keeping an eye on your massive re-write of FEpX.  Wow!   The new version is going to be quite an upgrade.  I look forward to trying it out.

**Reply by mkasemer** [reply]
Thanks Paul (@dplab)! I'm always a bit cautious when speaking for you :)
